### PR TITLE
handle negative values for shapes in PMMatrix

### PR DESCRIPTION
**Pull request by rakki-18**
This PR solves the issue - #205 partially.
Changes introduced with this PR :
- When negative values are given as input to the shapes of a PMMatrix, an AssertionFailure error is returned.

### CODE REVIEW

File: src/Math-Tests-Matrix/PMMatrixTest.class.st [hunk 1]
```diff
@@ -475,7 +475,10 @@ PMMatrixTest >> testMatrixInitialize [
 	| a |
 	a := PMMatrix new initialize: 2.
 	self assert: a numberOfRows equals: 2.
-	self assert: a numberOfColumns equals: 2
+	self assert: a numberOfColumns equals: 2.
+	
+	self should: [ PMMatrix new: -2 ] raise: AssertionFailure .
```
Comment (hemalvarambhia) [inline src/Math-Tests-Matrix/PMMatrixTest.class.st:480]: These, according to the project guidelines, should be a separate test method e.g. `testMatrixInitialisationWithNegativeDimensions` as this is an edge case.

File: src/Math-Matrix/PMMatrix.class.st [hunk 2]
```diff
@@ -520,7 +521,8 @@ PMMatrix >> initializeRows: anArrayOrVector [
 { #category : #initialization }
 PMMatrix >> initializeRows: rowsInteger columns: columnsInteger [
 	"Build empty components for a matrix."
-		
+	[ rowsInteger  isInteger and: [ rowsInteger   >= 0 ] ] assert.
+	[ columnsInteger  isInteger and: [ columnsInteger  >= 0 ] ] assert.
```
Comment (hemalvarambhia) [inline src/Math-Matrix/PMMatrix.class.st:525]: What failure message do clients get here? I'd it easy to debug from?

Comment (rakki-18) [reply src/Math-Matrix/PMMatrix.class.st:525]: It shows up as an assertion failure only.

Comment (hemalvarambhia) [reply src/Math-Matrix/PMMatrix.class.st:525]: Would you mind posting it here? It might be worth documenting that message in the test as an alternative.

Comment (rakki-18) [reply src/Math-Matrix/PMMatrix.class.st:525]: `AssertionFailure: Assertion failed`

Comment (hemalvarambhia) [reply src/Math-Matrix/PMMatrix.class.st:525]: Perhaps, then, we should be clearer e.g. 'Dimension of a matrix must be greater than zero'

File: src/Math-Matrix/PMMatrix.class.st [hunk 3]
```diff
@@ -508,6 +508,7 @@ PMMatrix >> hash [
 { #category : #initialization }
 PMMatrix >> initialize: anInteger [
 	"Build empty components for a square matrix. No check is made: components are assumed to be orgainized in rows."
+	[ anInteger isInteger and: [ anInteger  >= 0 ] ] assert.
```
Comment (hemalvarambhia) [inline src/Math-Matrix/PMMatrix.class.st:511]: Can we have a zero dimensional matrix? I don't think we do.

Comment (rakki-18) [reply src/Math-Matrix/PMMatrix.class.st:511]: I am not sure of this actually. But `Array2D` allows zero-dimensional matrix and we can create a `PMVector` of size 0. So, I felt we should allow zero-dimensional matrix in `PMMatrix`. What do you think?

Comment (hemalvarambhia) [reply src/Math-Matrix/PMMatrix.class.st:511]: This is a domain question that may be worth raising with Serge or in discord

Comment (hemalvarambhia) [reply src/Math-Matrix/PMMatrix.class.st:511]: Maybe the initialize method should be renamed after the discord conversation, and the comment for this method, which says clients get a square matrix

**Comment by SergeStinckwich** [comment]
Rename PMatrix>>initialize: as PMMatrix>>initializeSquare:
Use PMatrix>>rows:columns for PMMatrix>>initializeSquare: to avoid code duplication
Add tests to show these methods are working

File: src/Math-Matrix/PMMatrix.class.st [hunk 4]
```diff
@@ -105,7 +105,7 @@ PMMatrix class >> lupCRLCriticalDimension [
 PMMatrix class >> new: anInteger [
 	"Create an empty square matrix of dimension anInteger."
 
-	^ self new initialize: anInteger
+	^ self new initializeSquare: anInteger
```
Comment (hemalvarambhia) [inline src/Math-Matrix/PMMatrix.class.st:108]: `dimension` rather than `anInteger`

File: src/Math-Matrix/PMMatrix.class.st [hunk 5]
```diff
@@ -520,10 +514,17 @@ PMMatrix >> initializeRows: anArrayOrVector [
...
 PMMatrix >> initializeRows: rowsInteger columns: columnsInteger [
 	"Build empty components for a matrix."
-		
+	self assert: [ rowsInteger  isInteger and: [ rowsInteger   > 0 ] ] description: 'Row size of a matrix must be a positive integer'.
+	self assert: [ columnsInteger  isInteger and: [ columnsInteger  > 0 ] ]  description: 'Column size of a matrix must be a positive integer'.
 	rows := (1 to: rowsInteger) asPMVector collect: [ :each | PMVector new: columnsInteger ].
 ]
 
+{ #category : #initialization }
+PMMatrix >> initializeSquare: anInteger [
```
Comment (hemalvarambhia) [inline src/Math-Matrix/PMMatrix.class.st:523]: dimension rather than anInteger

**Comment by rakki-18** [comment]
@hemalvarambhia @SergeStinckwich, I have made the changes. Can you please check now?

File: src/Math-Matrix/PMMatrix.class.st [hunk 6]
```diff
@@ -102,10 +102,10 @@ PMMatrix class >> lupCRLCriticalDimension [
 ]
 
 { #category : #'instance creation' }
-PMMatrix class >> new: anInteger [
+PMMatrix class >> new: dimension [
```
Comment (hemalvarambhia) [inline src/Math-Matrix/PMMatrix.class.st:105]: Correct the variable name in the comment, if you don't mind.

**Review by hemalvarambhia — DISMISSED** [review]
Thank you for trying out our suggestions. Code looks a lot better.

File: src/Math-Matrix/PMMatrix.class.st (hunk 6 above)
Comment (rakki-18) [reply src/Math-Matrix/PMMatrix.class.st:105]: Yeah, done.

**Comment by hemalvarambhia** [comment]
@rakki-18 My colleague @SergeStinckwich suggested you might write programmer tests for the new messages & methods you introduced. I realised you are still working on that, so have dismissed my review. When those tests have been written, let us know on Discord.

**Comment by rakki-18** [comment]
@hemalvarambhia @SergeStinckwich The changes that I introduced:
- refactored and renamed `inititialize` method to `initializeSquare`. Test for this - [testMatrixInitializeSquare](https://github.com/PolyMathOrg/PolyMath/pull/206/files#diff-5b5b0273a566808328d352c6c2bb1e0e20dcc5ea78fb229bbb7513157131c27aR482)
- added a new assertion statement that prevents initialization of PMMatrix with non-positive dimensions. Test for this - [testMatrixInitializeWithNegativeDimensions](https://github.com/PolyMathOrg/PolyMath/pull/206/files#diff-5b5b0273a566808328d352c6c2bb1e0e20dcc5ea78fb229bbb7513157131c27aR474)
- Improved variable name in the PMMatrix `new:` method. Test for this - [testMatrixNew](https://github.com/PolyMathOrg/PolyMath/pull/206/files#diff-5b5b0273a566808328d352c6c2bb1e0e20dcc5ea78fb229bbb7513157131c27aR541) (Although, I am not sure if this test is required as it is a bit redundant to `testMatrixInitializeSquare`)

**Review by hemalvarambhia — APPROVED** [review]
